### PR TITLE
[SPARK-47702][CORE] Remove Shuffle service endpoint from the locations list when RDD block is removed form a node.

### PR DESCRIPTION
**Pull request by maheshk114**
### What changes were proposed in this pull request?

This PR fix a bug where the shuffle service end point is not removed when a RDD block is removed from a node. This is done by changing the call to reportBlockStatus in removeBlockInternal method. Currently StorageLevel.NONE is passed as storage level. This causes the driver to ignore the RDD in method updateBlockInfo as storageLevel.useDisk is set as false in StorageLevel.NONE.

### Why are the changes needed?

If the block location is not updated properly, then tasks fails with fetch failed exception. The tasks will try to read the RDD blocks from a node using external shuffle service. The read will fail, if the node is already decommissioned.

```
WARN BlockManager [Executor task launch worker for task 25.0 in stage 6.0 (TID 1567)]: Failed to fetch remote block rdd_5_25 from BlockManagerId(4, vm-92303839, 7337, None) (failed attempt 1)
org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:301)
	at org.apache.spark.network.BlockTransferService.fetchBlockSync(BlockTransferService.scala:103)
	at org.apache.spark.storage.BlockManager.fetchRemoteManagedBuffer(BlockManager.scala:1155)
	at org.apache.spark.storage.BlockManager.$anonfun$getRemoteBlock$8(BlockManager.scala:1099)
	at scala.Option.orElse(Option.scala:447)
	at org.apache.spark.storage.BlockManager.getRemoteBlock(BlockManager.scala:1099)
	at org.apache.spark.storage.BlockManager.getRemoteValues(BlockManager.scala:1045)
	at org.apache.spark.storage.BlockManager.get(BlockManager.scala:1264)
	at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:1326)
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a new UT.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by maheshk114** [comment]
@attilapiros Can you please take a look at this PR.

**Comment by attilapiros** [comment]
@maheshk114 I am sorry I missed this PR earlier. Let me try to reopen it.

**Comment by attilapiros** [comment]
retest

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -2112,8 +2112,10 @@ private[spark] class BlockManager(
       hasRemoveBlock = true
       if (tellMaster) {
         // Only update storage level from the captured block status before deleting, so that
-        // memory size and disk size are being kept for calculating delta.
-        reportBlockStatus(blockId, blockStatus.get.copy(storageLevel = StorageLevel.NONE))
+        // memory size and disk size are being kept for calculating delta. Reset the replica
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2115]:
```suggestion
        // memory size and disk size are being kept for calculating delta. Reset the replication
```

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -2112,8 +2112,10 @@ private[spark] class BlockManager(
...
+        // count 0 in storage level to notify that its a remove operation.
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2116]:
```suggestion
        // memory size and disk size are being kept for calculating delta. Reset the replication
        // factor to 0 to indicate its a remove operation.
```

**Comment by attilapiros** [comment]
@maheshk114 could you please open a new PR with the same content?

**Comment by mridulm** [comment]
I think it was the stale tag which caused it to get closed again @attilapiros. Removed it, let us see if it stays open :-)

**Pull request closed without merging**